In [1]:
# Install tmdbsimple (only need to run once)
!pip install tmdbsimple

In [2]:
import json
with open('/Users/jakec/.secret/tmdb_api.json', 'r') as f:
    login = json.load(f)
## Display the keys of the loaded dict
login.keys()

dict_keys(['client-id', 'api-key'])

In [3]:
import tmdbsimple as tmdb
tmdb.API_KEY =  login['client-id']

In [4]:
## make a movie object using the .Movies function from tmdb
movie = tmdb.Movies(603)
movie.info()
## movie objects have a .info dictionary 
#movie.info()


{'adult': False,
 'backdrop_path': '/giBJ0ezYNEobFBfB8H4JNTjmll.jpg',
 'belongs_to_collection': {'id': 2344,
  'name': 'The Matrix Collection',
  'poster_path': '/bV9qTVHTVf0gkW0j7p7M0ILD4pG.jpg',
  'backdrop_path': '/bRm2DEgUiYciDw3myHuYFInD7la.jpg'},
 'budget': 63000000,
 'genres': [{'id': 28, 'name': 'Action'},
  {'id': 878, 'name': 'Science Fiction'}],
 'homepage': 'http://www.warnerbros.com/matrix',
 'id': 603,
 'imdb_id': 'tt0133093',
 'original_language': 'en',
 'original_title': 'The Matrix',
 'overview': 'Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.',
 'popularity': 72.998,
 'poster_path': '/aOIuZAjPaRIE6CMzbazvcHuHXDc.jpg',
 'production_companies': [{'id': 79,
   'logo_path': '/tpFpsqbleCzEE2p5EgvUq6ozfCA.png',
   'name': 'Village Roadshow Pictures',
   'origin_country': 'US'},
  {'id': 372,
   'logo_path': None,
   'name': 'Groucho II Film 

In [9]:
def get_movie_with_rating(movie_id):
    movie=tmdb.Movies(movie_id)
    info = movie.info()
    releases = movie.releases()
    for c in releases['countries']:
        if c['iso_3166_1']=='US':
            info['certification'] = c['certification']
    return info

In [10]:
test = get_movie_with_rating("tt0848228") #put your function name here
test



{'adult': False,
 'backdrop_path': '/9BBTo63ANSmhC4e6r62OJFuK2GL.jpg',
 'belongs_to_collection': {'id': 86311,
  'name': 'The Avengers Collection',
  'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
  'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
 'budget': 220000000,
 'genres': [{'id': 878, 'name': 'Science Fiction'},
  {'id': 28, 'name': 'Action'},
  {'id': 12, 'name': 'Adventure'}],
 'homepage': 'https://www.marvel.com/movies/the-avengers',
 'id': 24428,
 'imdb_id': 'tt0848228',
 'original_language': 'en',
 'original_title': 'The Avengers',
 'overview': 'When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pull the world back from the brink of disaster. Spanning the globe, a daring recruitment effort begins!',
 'popularity': 109.732,
 'poster_path': '/RYMX2wcKCBAr24UyPD7xwmjaTn.jpg',
 'production_companies': [{'id': 420,
   'logo_path

In [11]:
## testing our function by looping through a list of ids
import pandas as pd
test_ids = ["tt0848228", "tt0115937","tt0848228","tt0332280"]
results = []
for movie_id in test_ids:
    movie_info = get_movie_with_rating(movie_id)
    results.append(movie_info)
    
    
pd.DataFrame(results)



HTTPError: 404 Client Error: Not Found for url: https://api.themoviedb.org/3/movie/tt0115937?api_key=39faaa4707ff475c076b889c3140c3a4

In [12]:
## testing our function by looping through a list of ids
import pandas as pd
test_ids = ["tt0848228", "tt0115937","tt0848228","tt0332280"]
results = []
for movie_id in test_ids:
    
    try:
        movie_info = get_movie_with_rating(movie_id)
        results.append(movie_info)
        
    except: 
        pass
    
pd.DataFrame(results)



adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   
1  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   
2  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   
1  {'id': 86311, 'name': 'The Avengers Collection...  220000000   
2                                               None   29000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
1  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0         https://www.marvel.com/movies/the-avengers  24428  tt0848228   
1         https://www.marvel.com/movies/the-avengers  24428  tt0848228   
2  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   
1                en   The Avengers  ...  1518815515      143   
2                en   The Notebook  ...   115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0                    Some assembly required.  The Avengers  False   
1                    Some assembly required.  The Avengers  False   
2  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0         7.708      28712         PG-13  
1         7.708      28712         PG-13  
2         7.879      10372         PG-13  

[3 rows x 26 columns]

In [13]:
## testing our function by looping through a list of ids
import pandas as pd
test_ids = ["tt0848228", "tt0115937","tt0848228","tt0332280"]
results = []
errors = []
for movie_id in test_ids:
    
    try:
        movie_info = get_movie_with_rating(movie_id)
        results.append(movie_info)
        
    except Exception as e: 
        errors.append([movie_id, e])
    
pd.DataFrame(results)



adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   
1  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   
2  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   
1  {'id': 86311, 'name': 'The Avengers Collection...  220000000   
2                                               None   29000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
1  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0         https://www.marvel.com/movies/the-avengers  24428  tt0848228   
1         https://www.marvel.com/movies/the-avengers  24428  tt0848228   
2  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   
1                en   The Avengers  ...  1518815515      143   
2                en   The Notebook  ...   115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0                    Some assembly required.  The Avengers  False   
1                    Some assembly required.  The Avengers  False   
2  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0         7.708      28712         PG-13  
1         7.708      28712         PG-13  
2         7.879      10372         PG-13  

[3 rows x 26 columns]

In [14]:
print(f"- Number of errors: {len(errors)}")
errors



- Number of errors: 1


[['tt0115937',
  requests.exceptions.HTTPError('404 Client Error: Not Found for url: https://api.themoviedb.org/3/movie/tt0115937?api_key=39faaa4707ff475c076b889c3140c3a4')]]

In [15]:
def write_json(new_data, filename): 
    """Appends a list of records (new_data) to a json file (filename). 
    Adapted from: https://www.geeksforgeeks.org/append-to-json-file-using-python/"""  
    
    with open(filename,'r+') as file:
        # First we load existing data into a dict.
        file_data = json.load(file)
        ## Choose extend or append
        if (type(new_data) == list) & (type(file_data) == list):
            file_data.extend(new_data)
        else:
             file_data.append(new_data)
        # Sets file's current position at offset.
        file.seek(0)
        # convert back to json.
        json.dump(file_data, file)


In [16]:
import pandas as pd
# Load in the dataframe from project part 1 as basics:
basics = pd.read_csv("Data/title_basics.csv.gz", low_memory = False)

In [17]:
YEARS_TO_GET = [2000,2001]

errors = [ ]



In [21]:
from tqdm.notebook import tqdm_notebook
FOLDER = "Data/"
import os
# Start of OUTER loop
for YEAR in tqdm_notebook(YEARS_TO_GET, desc='YEARS', position=0):
    #Defining the JSON file to store results for year
    JSON_FILE = f'{FOLDER}tmdb_api_results_{YEAR}.json'
    # Check if file exists
    file_exists = os.path.isfile(JSON_FILE)
    # If it does not exist: create it
    if file_exists == False:
    # save an empty dict with just "imdb_id" to the new json file.
        with open(JSON_FILE,'w') as f:
            json.dump([{'imdb_id':0}],f)
    #Saving new year as the current df
    df = basics.loc[ basics['startYear']==YEAR].copy()
    # saving movie ids to list
    movie_ids = df['tconst'].copy()
    # Load existing data from json into a dataframe called "previous_df"
    previous_df = pd.read_json(JSON_FILE)
    # filter out any ids that are already in the JSON_FILE
    movie_ids_to_get = movie_ids[~movie_ids.isin(previous_df['imdb_id'])]
        #Get index and movie id from list
    # INNER Loop
    for movie_id in tqdm_notebook(movie_ids_to_get,
                                  desc=f'Movies from {YEAR}',
                                  position=1,
                                  leave=True):
        try:
            # Retrieve then data for the movie id
            temp = get_movie_with_rating(movie_id)  
            # Append/extend results to existing file using a pre-made function
            write_json(temp,JSON_FILE)
            # Short 20 ms sleep to prevent overwhelming server
            time.sleep(0.02)
            
        except Exception as e:
            errors.append([movie_id, e])
    final_year_df = pd.read_json(JSON_FILE)
    final_year_df.to_csv(f"{FOLDER}final_tmdb_data_{YEAR}.csv.gz", compression="gzip", index=False)

print(f"- Total errors: {len(errors)}")


YEARS:   0%|          | 0/2 [00:00<?, ?it/s]

Movies from 2000:   0%|          | 0/1448 [00:00<?, ?it/s]

Movies from 2001:   0%|          | 0/1569 [00:00<?, ?it/s]

- Total errors: 3017


In [24]:
data2000 = pd.read_csv("Data/final_tmdb_data_2000.csv.gz", low_memory = False)
data2001 = pd.read_csv("Data/final_tmdb_data_2001.csv.gz", low_memory = False)

In [26]:
datacombined=pd.concat([data2000, data2001])

In [27]:
datacombined

imdb_id  adult                     backdrop_path  \
0             0    NaN                               NaN   
1     tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg   
2     tt0113092    0.0                               NaN   
3     tt0116391    0.0                               NaN   
4     tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   
...         ...    ...                               ...   
1331  tt7797790    0.0                               NaN   
1332  tt8665056    0.0                               NaN   
1333  tt8795764    0.0                               NaN   
1334  tt9071078    0.0                               NaN   
1335  tt9212730    0.0                               NaN   

                                  belongs_to_collection      budget  \
0                                                   NaN         NaN   
1                                                   NaN  10000000.0   
2                                                   NaN         0.0   
3                                                   NaN         0.0   
4     {'id': 1131062, 'name': 'Wong Kar-Wai’s Love T...    150000.0   
...                                                 ...         ...   
1331                                                NaN         0.0   
1332                                                NaN         0.0   
1333                                                NaN         0.0   
1334                                                NaN         0.0   
1335                                                NaN         0.0   

                                                 genres  \
0                                                   NaN   
1     [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
2              [{'id': 878, 'name': 'Science Fiction'}]   
3     [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
4     [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
...                                                 ...   
1331                     [{'id': 27, 'name': 'Horror'}]   
1332                    [{'id': 37, 'name': 'Western'}]   
1333                     [{'id': 27, 'name': 'Horror'}]   
1334                     [{'id': 28, 'name': 'Action'}]   
1335  [{'id': 27, 'name': 'Horror'}, {'id': 53, 'nam...   

                                               homepage        id  \
0                                                   NaN       NaN   
1                                                   NaN   62127.0   
2                                                   NaN  110977.0   
3                                                   NaN  442869.0   
4                                                   NaN     843.0   
...                                                 ...       ...   
1331                                                NaN  956219.0   
1332  http://skeletoncreekproductions.com/p-movie-br...  885436.0   
1333  https://www.utahwolf.com/films/coming-soon-new...  871624.0   
1334                                                NaN  201706.0   
1335                                                NaN  382630.0   

     original_language                       original_title  ...     revenue  \
0                  NaN                                  NaN  ...         NaN   
1                   en                      The Fantasticks  ...         0.0   
2                   en                        For the Cause  ...         0.0   
3                   hi                                 Gang  ...         0.0   
4                   cn                                 花樣年華  ...  14204632.0   
...                ...                                  ...  ...         ...   
1331                en  Edmund Kemper Part 3: La mort sévit  ...         0.0   
1332                en                 Guns Along The Bravo  ...         0.0   
1333                en                            New Breed  ...         0.0   
1334                cn                                 致命密函  ...         0.0   
1335            

How many movies had at least some valid financial information (values > 0 for budget OR revenue)?

In [30]:
((datacombined['budget'] > 0) ^ (datacombined['revenue'] > 0)).sum()

284

284 movies.

How many movies are there in each of the certification categories (G/PG/PG-13/R)?


In [31]:
datacombined['certification'].value_counts()

R          467
PG-13      182
NR          71
PG          63
G           25
NC-17        6
Unrated      1
Name: certification, dtype: int64

What is the average revenue per certification category?


In [32]:
datacombined.groupby('certification')['revenue'].mean()

certification
G          7.041886e+07
NC-17      0.000000e+00
NR         2.295880e+06
PG         6.148838e+07
PG-13      7.146544e+07
R          1.612925e+07
Unrated    0.000000e+00
Name: revenue, dtype: float64

What is the average budget per certification category?


In [33]:
datacombined.groupby('certification')['budget'].mean()

certification
G          2.288000e+07
NC-17      0.000000e+00
NR         1.509015e+06
PG         2.490472e+07
PG-13      3.094592e+07
R          9.663393e+06
Unrated    0.000000e+00
Name: budget, dtype: float64